Sources/SemanticIndex/UpdateIndexStoreTaskDescription.swift

```diff
@@ -0,0 +1,311 @@
+        .joined(separator: ", ")
+      logger.log(
+        "Starting updating index store with priority \(Task.currentPriority.rawValue, privacy: .public): \(filesToIndexDescription)"
+      )
+      let filesToIndex = filesToIndex.sorted(by: { $0.stringValue < $1.stringValue })
+      // TODO (indexing): Once swiftc supports it, we should group files by target and index files within the same
+      // target together in one swiftc invocation.
+      for file in filesToIndex {
+        await updateIndexStoreForSingleFile(file)
+      }
+      logger.log(
+        "Finished updating index store in \(Date().timeIntervalSince(startDate) * 1000, privacy: .public)ms: \(filesToIndexDescription)"
+      )
+    }
+  }
+
+  public func dependencies(
+    to currentlyExecutingTasks: [UpdateIndexStoreTaskDescription]
+  ) -> [TaskDependencyAction<UpdateIndexStoreTaskDescription>] {
+    return currentlyExecutingTasks.compactMap { (other) -> TaskDependencyAction<UpdateIndexStoreTaskDescription>? in
+      guard !other.filesToIndex.intersection(filesToIndex).isEmpty else {
+        // Disjoint sets of files can be indexed concurrently.
+        return nil
+      }
+      if self.filesToIndex.count < other.filesToIndex.count {
+        // If there is an index operation with more files already running, suspend it.
+        // The most common use case for this is if we schedule an entire target to be indexed in the background and then
+        // need a single file indexed for use interaction. We should suspend the target-wide indexing and just index
+        // the current file to get index data for it ASAP.
+        return .cancelAndRescheduleDependency(other)
+      } else {
+        return .waitAndElevatePriorityOfDependency(other)
+      }
+    }
+  }
+
+  private func updateIndexStoreForSingleFile(_ uri: DocumentURI) async {
+    guard let url = uri.fileURL else {
+      // The URI is not a file, so there's nothing we can index.
+      return
+    }
+    guard !index.hasUpToDateUnit(for: url) else {
+      // We consider a file's index up-to-date if we have any up-to-date unit. Changing build settings does not
+      // invalidate the up-to-date status of the index.
+      return
+    }
+    guard let language = await buildSystemManager.defaultLanguage(for: uri) else {
+      logger.error("Not indexing \(uri.forLogging) because its language could not be determined")
+      return
+    }
+    let buildSettings = await buildSystemManager.buildSettingsInferredFromMainFile(
+      for: uri,
+      language: language,
+      logBuildSettings: false
+    )
+    guard let buildSettings else {
+      logger.error("Not indexing \(uri.forLogging) because it has no compiler arguments")
+      return
+    }
+    guard !buildSettings.isFallback else {
+      // Only index with real build settings. Indexing with fallback arguments could result in worse results than not
```

Comment:
When could this actually happen? We're not using index-while-building here, so it seems like an odd case. Presumably having *something* is better than nothing, but even if you added a "do we have anything in the index" check, that would still be odd as then we'd index it the first time and not again after that.